### PR TITLE
Remove not needed column permutations

### PR DESCRIPTION
**Pull request by rasolca**
Based on #843

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h [hunk 1]
```diff
@@ -43,6 +43,31 @@ void initIndex(const SizeType i_begin, const SizeType i_end, Matrix<SizeType, D>
...
+    SizeType* index_ptr = index[0].ptr(zero_idx);
+
+    auto begin_it = index_ptr;
+    auto end_it = index_ptr + n;
+    std::for_each(begin_it, end_it, [val](SizeType& i) { i += val; });
+  };
+
+  TileCollector tc{i_begin, i_end};
+
+  auto sender = ex::when_all(ex::when_all_vector(tc.readwrite<SizeType, Device::CPU>(index)));
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/index_manipulation.h:65]:
```suggestion
  auto sender = ex::when_all_vector(tc.readwrite<SizeType, Device::CPU>(index));
```

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 2]
```diff
@@ -41,26 +41,28 @@ namespace dlaf::eigensolver::internal {
 
 // Auxiliary matrix and vectors used for the D&C algorithm
 //
-// - mat1 and mat2: (matrices)
-//   Extra workspace for Q1', Q2' and U1' if n2 > n1 or U2' if n1 >= n2 which are packed as follows:
+// - e0: (matrix)
+//     In: Holds the eigenvectors of the two subproblems TODO in ascending order.
+//     Out: Holds the eigenvectors of the merged problems TODO in ascending order.
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:46]: `TODO`?

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 3]
```diff
@@ -41,26 +41,28 @@ namespace dlaf::eigensolver::internal {
...
-//   ├──────┤      │   │
-//   │      └──────┘   │
-//   │                 │
-//   ├────────────┐    │
-//   │     U1'    │    │
-//   │            │    │
-//   └────────────┴────┘
+// - d0: (vector)
+//     In: Holds the eigenvalues of the two subproblems TODO in ascending order.
+//     Out: Holds the eigenvalues of the merged problems TODO in ascending order.
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:54]: `TODO`?

**Review by msimberg — CHANGES_REQUESTED** [review]

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 3 above)
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: https://github.com/eth-cscs/DLA-Future/pull/858/commits/79e3475a7b2a6f4fb77c1af09909a19d7fb1ec00

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 2 above)
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: Sorry, I added it as the documentation was not holding anymore, and forgot to update it...
https://github.com/eth-cscs/DLA-Future/pull/858/commits/79e3475a7b2a6f4fb77c1af09909a19d7fb1ec00

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h (hunk 1 above)
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/index_manipulation.h]: https://github.com/eth-cscs/DLA-Future/pull/858/commits/79e3475a7b2a6f4fb77c1af09909a19d7fb1ec00

**Review by msimberg — APPROVED** [review]

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/eigensolver/tridiag_solver/impl.h [hunk 4]
```diff
@@ -188,51 +190,76 @@ void offloadDiagonal(Matrix<const T, Device::CPU>& tridiag, Matrix<T, D>& evals)
...
   // Each triad represents two subproblems to be merged
   for (auto [i_begin, i_split, i_end] : generateSubproblemIndices(distr.nrTiles().rows())) {
-    mergeSubproblems<B>(i_begin, i_split, i_end, offdiag_vals[to_sizet(i_split - 1)], ws, ws_h, ws_hm,
-                        evals, evecs);
+    mergeSubproblems<B>(i_begin, i_split, i_end, offdiag_vals[to_sizet(i_split - 1)], ws, ws_h, ws_hm);
   }
 
-  copy({0, 0}, evals.distribution().localNrTiles(), ws_hm.evals, evals);
+  const SizeType n = evecs.nrTiles().rows();
+  copy({0, 0}, ws.i2.distribution().localNrTiles(), ws_hm.i2, ws.i2);
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/impl.h:256]: I'm missing something: why do we have to specify begin/size? aren't they distributed the same way?

```suggestion
  copy(ws_hm.i2, ws.i2);
```

There are different places where this is done, for which the question might be similar.

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 5]
```diff
@@ -41,29 +41,35 @@ namespace dlaf::eigensolver::internal {
...
+// - d0: (vector)
+//     In: Holds the eigenvalues of the two subproblems (in ascending order if permuted with i1).
+//     Out: Holds the eigenvalues of the merged problems (in ascending order if permuted with i1).
+// - d1: (vector)
+//     Holds the values of the deflated diagonal sorted in ascending order
+// - z0: (vector)
+//     Holds the values of Cuppen's rank-1 vector
+//     Also used as temporary workspace
+// - z1: (vector)
+//     Holds the values of the rank-1 update vector sorted as d1
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:61]: ?
```suggestion
//     Holds the values of the rank-1 update vector (same order as d1)
```

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 6]
```diff
@@ -41,29 +41,35 @@ namespace dlaf::eigensolver::internal {
 
 // Auxiliary matrix and vectors used for the D&C algorithm
 //
-// - mat1 and mat2: (matrices)
-//   Extra workspace for Q1', Q2' and U1' if n2 > n1 or U2' if n1 >= n2 which are packed as follows:
+// - e0: (matrix)
+//     In: Holds the eigenvectors of the two subproblems (same order as d0).
+//     Out: Holds the eigenvectors of the merged problems (same order as d0).
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:46]: ?
```suggestion
//     Out: Holds the eigenvectors of the merged problem (same order as d0).
```

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 7]
```diff
@@ -41,29 +41,35 @@ namespace dlaf::eigensolver::internal {
...
-//   ├──────┤      │   │
-//   │      └──────┘   │
-//   │                 │
-//   ├────────────┐    │
-//   │     U1'    │    │
-//   │            │    │
-//   └────────────┴────┘
+// - d0: (vector)
+//     In: Holds the eigenvalues of the two subproblems (in ascending order if permuted with i1).
+//     Out: Holds the eigenvalues of the merged problems (in ascending order if permuted with i1).
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:54]: ?
```suggestion
//     Out: Holds the eigenvalues of the merged problem (in ascending order if permuted with i1).
```

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 8]
```diff
@@ -1039,81 +1059,76 @@ void mergeDistSubproblems(comm::CommunicatorGrid grid,
...
+  applyIndex(i_begin, i_end, ws_h.i3, ws_h.d0, ws_hm.d1);
+  applyIndex(i_begin, i_end, ws_h.i3, ws_hm.z0, ws_hm.z1);
+  copy(idx_begin_tiles_vec, sz_tiles_vec, ws_hm.d1, ws_h.d0);
 
   //
   //    i3 (in)  : initial <--- deflated
   //    i2 (out) : initial ---> deflated
   //
   invertIndex(i_begin, i_end, ws_h.i3, ws_hm.i2);
+  //////
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:1085]: ?
```suggestion
```

**Review by albestro — APPROVED** [review]
LGTM

just a bunch of minor comments.

**Pull request merged**